Userland/Libraries/LibC/arch/x86_64/setjmp.S

```diff
@@ -10,7 +10,9 @@
 // /!\ Read setjmp.h before modifying this file!
 //
 
+.global _setjmp
 .global setjmp
+_setjmp:
```

Comment:
According to the OpenBSD man pages, these functions are in fact not equivalent: 

https://man.openbsd.org/setjmp.3

> The [sigsetjmp](https://man.openbsd.org/setjmp.3#sigsetjmp~3)()/[siglongjmp](https://man.openbsd.org/setjmp.3#siglongjmp~2)() function pairs save and restore the signal mask if the argument savemask is non-zero. Otherwise, only the register set and the stack are saved.

> In other words, [setjmp](https://man.openbsd.org/setjmp.3#setjmp~2)()/longjmp() are functionally equivalent to sigsetjmp()/siglongjmp() when sigsetjmp() is called with a non-zero savemask argument. Conversely, _setjmp()/_longjmp() are functionally equivalent to sigsetjmp()/siglongjmp() when sigsetjmp() is called with a zero-value savemask.

... but then it goes on to say that

> Historically, on AT&T System V UNIX, the setjmp()/longjmp() functions have been equivalent to the BSD _setjmp()/_longjmp() functions and do not restore the signal mask. Because of this discrepancy, the sigsetjmp()/siglongjmp() interfaces should be used if portability is desired.

... and the linux man pages say

```
       POSIX does not specify whether setjmp() will save the signal mask
       (to be later restored during longjmp()).  In System V it will
       not.  In 4.3BSD it will, and there is a function _setjmp() that
       will not.  The behavior under Linux depends on the glibc version
       and the setting of feature test macros.  On Linux with glibc
       versions before 2.19, setjmp() follows the System V behavior by
       default, but the BSD behavior is provided if the _BSD_SOURCE
       feature test macro is explicitly defined and none of
       _POSIX_SOURCE, _POSIX_C_SOURCE, _XOPEN_SOURCE, _GNU_SOURCE, or
       _SVID_SOURCE is defined.  Since glibc 2.19, <setjmp.h> exposes
       only the System V version of setjmp().  Programs that need the
       BSD semantics should replace calls to setjmp() with calls to
       sigsetjmp() with a nonzero savesigs argument.
```

So, with all that said, I think we probably want to either 1) emulate BSD behavior or 2) do what glibc 2.19 does and not have `_setjmp` at all.

Reply:
I'll tag @bgianfo for thoughts here as he added the initial implementation of _setjmp and _longjmp for use with syzkaller.

Reply:
> According to the OpenBSD man pages, these functions are in fact not equivalent:

Yeah... its complicated.

Here is how its done in existing systems:

glibc calls sigsetjmp with a non-zero value for the non-underscored and a zero value for the underscored variant
[setjmp](https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/x86_64/bsd-setjmp.S;h=47b662a2812a1ba3d9946f0668e7bd7e5402404c;hb=HEAD#l28)
[_setjmp](https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/x86_64/bsd-_setjmp.S;h=fb58b273db86645196bd4b4b1d522a77e6e2bba2;hb=HEAD#l28)

musl implements both as the same
[setjmp](https://github.com/bminor/musl/blob/cfdfd5ea3ce14c6abf7fb22a531f3d99518b5a1b/src/setjmp/x86_64/setjmp.s)


Serenity implements them the same for i386
[setjmp](https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibC/arch/i386/setjmp.S#L19)


MinGW is complicated and uses one of the many setjmp variants Windows exports
but on x86_64 setjmp appears ends up as _setjmp
[setjmp](https://github.com/mingw-w64/mingw-w64/blob/826706cbf2b01e799f495bf57d5e681da74dad37/mingw-w64-headers/crt/setjmp.h#L242)

FreeBSD implements setjmp and _setjmp separately
[setjmp](https://github.com/freebsd/freebsd-src/blob/5d56371c70d785c222ff086ba8a83b408111acbf/lib/libc/amd64/gen/setjmp.S#L51)
[_setjmp](https://github.com/freebsd/freebsd-src/blob/main/lib/libc/amd64/gen/_setjmp.S#L49)


Personally I think the glibc approach of having both variants call `sigsetjmp` with the correct val may be the best solution to this.

> So, with all that said, I think we probably want to either 1) emulate BSD behavior

Either way, some projects like Lua put the use of `_{long,set}jmp` behind a `USE_POSIX` flag, even if its not Posix.
I think the underscore variants should be defined and declared, at least for compatibility purposes.

> or 2) do what glibc >=2.19 does and not have `_setjmp` at all.

I need some clarification on this.
Looking through the glibc source code, it still appears to declare and define `_{set,long}jmp` (even if its just a call to sigsetjmp)

https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/x86_64/bsd-_setjmp.S;h=fb58b273db86645196bd4b4b1d522a77e6e2bba2;hb=HEAD#l28

Reply:
well that's super interesting. I checked the glibc public header too and both are exposed to all users.

https://man7.org/linux/man-pages/man3/setjmp.3.html

I wonder why the man page here is lying :thinking: .

Reply:
that man page could definietely benefit from mentioning `_setjmp` but for the time being https://man7.org/linux/man-pages/man0/setjmp.h.0p.html exists and documents the existence of `_setjmp`

Reply:
Eh, we can debate whether our current behavior is correct or not and maybe call sigsetjmp with 0 or 1 from each method in another PR, for now it seems like getting the behavior equivalent on all platforms is still valuable. So i'll merge this as-is.

The fact that sigsetjmp will save the procmask but longjmp never restores it feels very strange to me as well, but, yah, out of scope probably.